### DISCUSSION TITLE
Question about the TSN showcases (and similar concepts)

### DISCUSSION THREAD

**Discussion by anoikob** [omnetpp/omnetpp#1220] [Q&A]
Hey there. I am checking out the TSN showcases, and I'm trying to add a tsnSwitch module between the two devices in the FramePreemption showcase. I have added it, connected it to the devices, but when I run the simulation, the messages generated by the  device on the left never leave the switch to go to the device on the right. I know I am supposed to edit some routing stuff. And here is the problem:

In the showcase, I'm only presented with a .ned file and a .ini file, not a source file. I know that .ned files are only used for network description aka topology, connections and module placement, and that the source files are used for the actual operation of the devices. What do I have to edit/create so that the switch not only sees the 2 devices (yes I enabled frame preemption on the switch), but also passes messages from one device to the other?

I'm not straight up asking for code, just for directions. Thanks a lot!

**Comment by levy** [comment]
The packets should be sent out from the switch. I don't know if you are using ARP or the MAC forwarding table is pre-filled. You can see where and why the packets are stuck or dropped in the log window of Qtenv. You can also post your NED/INI file here and let us see if there's anything suspicious.

**Reply by anoikob** [reply]
This is the .ned source file 
![image](https://github.com/user-attachments/assets/71208423-5cf9-4a30-af88-df3bbf0307fd)

Which makes this network, visually
![image](https://github.com/user-attachments/assets/9bc4bb49-a9e7-49af-a663-001a3466849f)

The .ini file is this (won't let me upload it outright): 
`[General]
network = FramePreemptionShowcase
sim-time-limit = 1s
#abstract-config = true (requires omnet 7)

*.*.ethernet.typename = "EthernetLayer"
*.host*.eth[0].typename = "LayeredEthernetInterface"

*.host*.eth[0].bitrate = 100Mbps

**.recordPcap = true
**.dumpProtocols = "ethernetphy"

**.crcMode = "computed"
**.fcsMode = "computed"

**.arp.typename = "GlobalArp"

# host1
*.host1.numApps = 2
*.host1.app[*].typename = "UdpSourceApp"
*.host1.app[0].source.packetNameFormat = "background-%c"
*.host1.app[1].source.packetNameFormat = "ts-%c"
*.host1.app[*].tagger.typename = "PacketTagger"
*.host1.app[0].tagger.vlanId = 1
*.host1.app[1].tagger.vlanId = 0
*.host1.app[*].io.destAddress = "host2"
*.host1.app[0].io.destPort = 1000
*.host1.app[1].io.destPort = 1001

# background ~96Mbps
*.host1.app[0].source.packetLength = 1200B
*.host1.app[0].source.productionInterval = truncnormal(100us,50us)
*.host1.app[0].source.initialProductionOffset = 5us

# high-ts ~9.6Mbps
*.host1.app[1].source.packetLength = 1200B
*.host1.app[1].source.productionInterval = truncnormal(1ms,500us)

# host2
*.host2.numApps = 2
*.host2.app[*].typename = "UdpSinkApp"
*.host2.app[0].io.localPort = 1000
*.host2.app[1].io.localPort = 1001

[Config FifoQueueing]
description = "High and low priority frames are transmitted in first-in first-out order"

**.macLayer.queue.packetCapacity = 4
**.macLayer.queue.typename = "DropTailQueue"

[Config PriorityQueueing]
description = "High priority frames are transmitted before low priority frames"

**.macLayer.queue.typename = "PriorityQueue"
**.macLayer.queue.numQueues = 2
**.macLayer.queue.queue[*].packetCapacity = 4

**.macLayer.queue.queue*.typename = "DropTailQueue"
**.macLayer.queue.classifier.classifierClass = "inet::PacketVlanReqClassifier"

[Config FramePreemption]
description = "Transmission of low priority frames are preempted by high priority frames"

*.host*.eth[0].macLayer.typename = "EthernetPreemptingMacLayer"
*.host*.eth[0].phyLayer.typename = "EthernetPreemptingPhyLayer"

**.macLayer.*.queue.packetCapacity = 4
**.macLayer.*.queue.typename = "DropTailQueue"

[Config RealisticBase]
#abstract-config = true (requires omnet 7)

# background
*.host1.app[0].source.packetLength = 1200B
*.host1.app[0].source.productionInterval = truncnormal(100us,50us)

# time-sensitive
*.host1.app[1].source.packetLength = 120B
*.host1.app[1].source.productionInterval = truncnormal(10ms,5ms)

[Config RealisticFifoQueueing]
extends = RealisticBase, FifoQueueing

[Config RealisticPriorityQueueing]
extends = RealisticBase, PriorityQueueing

[Config RealisticFramePreemption]
extends = RealisticBase, FramePreemption
`
As you can see, the tsnSwitch does not appear in the .ini file, but that's because I'm not yet familiar with the correct syntax, so I chose to see how would the network work as is, if I just dropped a TSN switch and enabled its frame preemption. I'm sure I need to insert it and give it some addresses or something, though I don't see something similar when it comes to the 2 hosts (maybe because they don't need explicit addresses because they are only talking to each other). Thanks for the reply!

edit: As a bonus, I managed to insert the tsnSwitch into the ini file, and set its speed, using the syntax the hosts had. It worked, cause it did not require me to set the speed during the initialization in the Qtenv  window
![image](https://github.com/user-attachments/assets/81290cb0-9f60-4b3a-ac40-e4be60d544b4)
